src/tape/icp.jl

```diff
@@ -0,0 +1,175 @@
+# References
+# [1] ReverseDiff.jl - Jarrett Revels
+# [2] Applied Interval Analysis - Luc Jaulin, Michel Kieffer, Olivier Didrit and Eric Walter
+
+"""
+Function to apply the forward-reverse contractor on `input` for `function` f. Returns the contracted array, not modifying `input`.
+
+`function` should currently be a `n to 1` function.
+
+`input` must be a `AbstractVector` of `Interval`s.
+
+`contraint` must be an `Interval`.
+
+`tape` is an optional argument, and should be a `Tape` of `ScalarInstruction`s, which can be pre-recorded to avoid recording redundantly. To record -
+
+`tape = IntervalConstraintProgramming.Tape(f, X, constraint)`
+
+Usage:
+```
+julia> f(X) = X[1]^2 + X[2]^2
+f (generic function with 1 method)
+
+julia> X = [-100..100, -100..100]
+2-element Array{IntervalArithmetic.Interval{Float64},1}:
+ [-100, 100]
+ [-100, 100]
+
+julia> constraint = -∞..1
+[-∞, 1]
+
+julia> icp(f, X, constraint)
+2-element Array{IntervalArithmetic.Interval{Float64},1}:
+ [-1, 1] [-1, 1]
+```
+
+"""
+function icp(f::Function, input::AbstractArray, constraint::Interval)
+    tape = Tape(f, input, constraint)
+    reverse_pass!(tape.tape, tape.input.value)
+    return tape.input.value
+end
+
+function icp(f::Function, input::AbstractArray, constraint::Interval, tape::AbstractTape)
+    reverse_pass!(tape.tape, tape.input.value)
+    return tape.input.value
+end
+
+function icp(f::Function, input::IntervalBox, constraint::Interval)
+    tape = Tape(f, input.v, constraint)
+    reverse_pass!(tape.tape, tape.input.value)
+    return IntervalBox(tape.input.value)
+end
+
+function icp(f::Function, input::IntervalBox, constraint::Interval, tape::AbstractTape)
+    reverse_pass!(tape.tape, tape.input.value)
+    return IntervalBox(tape.input.value)
+end
+
+"""
+Function to apply the forward-reverse contractor on `input` for `function` f. Modifies `input` to contain the contracted array of `Interval`s.
+
+`function` should currently be a `n to 1` function.
+
+`input` must be a `AbstractVector` of `Interval`s.
+
+`contraint` must be an `Interval`.
+
+`tape` is an optional argument, and should be a `Tape` of `ScalarInstruction`s, which can be pre-recorded to avoid recording redundantly. To record -
+
+`tape = IntervalConstraintProgramming.Tape(f, X, constraint)`
+
+Usage:
+```
+julia> f(X) = X[1]^2 + X[2]^2
+f (generic function with 1 method)
+
+julia> X = [-100..100, -100..100]
+2-element Array{IntervalArithmetic.Interval{Float64},1}:
+ [-100, 100]
+ [-100, 100]
+
+julia> constraint = -∞..1
+[-∞, 1]
+
+julia> icp!(f, X, constraint)
+2-element Array{IntervalArithmetic.Interval{Float64},1}:
+ [-1, 1] [-1, 1]
+```
+
+"""
+
+function icp!(f::Function, input::AbstractArray, constraint::Interval)
```

Comment:
The convention is that the first argument is modified, so I think we should reverse the order of the first two arguments.

Reply:
Actually, if the mutating function accepts a function argument, then that usually comes first, e.g.

```julia
help?> map!
search: map! asyncmap! map mapfoldr mapfoldl mapslices mapreduce mapreducedim asyncmap macroexpand @macroexpand

  map!(function, destination, collection...)

  Like map, but stores the result in destination rather than a new collection. destination must be at least as large
  as the first collection.
```

with the reason being that it allows use of the convenient `do` notation.